src/v/datalake/proto_to_arrow_struct.cc

```diff
@@ -0,0 +1,139 @@
+    }
+    for (size_t child_idx = 0; child_idx < _child_arrays.size(); child_idx++) {
+        _arrow_status = _child_arrays[child_idx]->add_value(
+          msg, int(child_idx));
+        if (!_arrow_status.ok()) {
+            return _arrow_status;
+        }
+    }
+    _arrow_status = _builder->Append();
+
+    return _arrow_status;
+}
+arrow::Status datalake::detail::proto_to_arrow_struct::add_value(
+  const google::protobuf::Message* msg, int field_idx) {
+    if (!_arrow_status.ok()) {
+        return _arrow_status;
+    }
+    auto desc = msg->GetDescriptor()->field(field_idx);
+    auto child_message = &msg->GetReflection()->GetMessage(*msg, desc);
+    return add_top_level_message(child_message);
+}
+datalake::detail::proto_to_arrow_struct::proto_to_arrow_struct(
+  const google::protobuf::Descriptor* message_descriptor) {
+    using namespace detail;
+    namespace pb = google::protobuf;
+
+    // Set up child arrays
+    _child_arrays.reserve(message_descriptor->field_count());
+    for (int field_idx = 0; field_idx < message_descriptor->field_count();
+         field_idx++) {
+        auto desc = message_descriptor->field(field_idx);
+        switch (desc->cpp_type()) {
+        case pb::FieldDescriptor::CPPTYPE_INT32:
+            _child_arrays.push_back(
+              std::make_unique<proto_to_arrow_scalar<arrow::Int32Type>>());
+            break;
+        case pb::FieldDescriptor::CPPTYPE_INT64:
+            _child_arrays.push_back(
+              std::make_unique<proto_to_arrow_scalar<arrow::Int64Type>>());
+            break;
+        case pb::FieldDescriptor::CPPTYPE_BOOL:
+            _child_arrays.push_back(
+              std::make_unique<proto_to_arrow_scalar<arrow::BooleanType>>());
+            break;
+        case pb::FieldDescriptor::CPPTYPE_FLOAT:
+            _child_arrays.push_back(
+              std::make_unique<proto_to_arrow_scalar<arrow::FloatType>>());
+            break;
+        case pb::FieldDescriptor::CPPTYPE_DOUBLE:
+            _child_arrays.push_back(
+              std::make_unique<proto_to_arrow_scalar<arrow::DoubleType>>());
+            break;
+        case pb::FieldDescriptor::CPPTYPE_STRING:
+            _child_arrays.push_back(
+              std::make_unique<proto_to_arrow_scalar<arrow::StringType>>());
+            break;
+        case pb::FieldDescriptor::CPPTYPE_MESSAGE: {
+            auto field_message_descriptor = desc->message_type();
+            if (field_message_descriptor == nullptr) {
+                throw datalake::initialization_error(
+                  std::string("Can't find schema for nested type : ")
+                  + desc->cpp_type_name());
+            }
+            _child_arrays.push_back(std::make_unique<proto_to_arrow_struct>(
+              field_message_descriptor));
+        } break;
+        default:
+            throw datalake::initialization_error(
+              fmt::format("Unknown type: {}", desc->cpp_type_name()));
+        }
```

Comment:
nit: what fields aren't covered? IMO if it's not a ton, it's a bit less error prone to explicitly enumerate the non-supported ones here.